File: wiki/eject.py
"""eject: Replace symlinks in the target repo with real copies of the wiki docs.

Use this to detach one or all paths from the wiki. The wiki doc is preserved;
only the symlink in the target repo is replaced with a real file.

After ejecting, remove the path from schema.yaml if you no longer want the wiki
to manage it — or re-run push to restore the symlink.
"""

import shutil
import subprocess
from pathlib import Path

from .lib import (
    DOCS_ROOT, WIKI_ROOT, clear_conflict_log_for, clear_flag, doc_path,
    get_doc_filename, get_repo_path, load_conflict_log, load_schema, schema_paths,
    strip_wiki_header, symlink_path, walk_schema,
)


def _no_skip_worktree(repo: Path, rel_file: str):
    subprocess.run(
        ["git", "-C", str(repo), "update-index", "--no-skip-worktree", rel_file],
        capture_output=True,
    )


def run_eject(scope: str | None = None):
    repo = get_repo_path()
    schema = load_schema()
    nodes = walk_schema(schema)

    if scope is not None:
        nodes = [(rp, meta) for rp, meta in nodes if rp == scope]
        if not nodes:
            s_paths = schema_paths(schema)
            raise SystemExit(
                f"'{scope}' is not a managed doc path.\n"
                f"Managed paths: {', '.join(s_paths) or '(none)'}"
            )

    fn = get_doc_filename()
    symlinks_dir = repo / ".agent-wiki" / "symlinks"
    local_edits_dir = WIKI_ROOT / "logs" / "local-edits"
    ejected_paths = []

    for rel_path, _ in nodes:
        link = symlink_path(repo, rel_path)
        wiki_doc = doc_path(rel_path)
        display = f"docs/{rel_path}/{fn}" if rel_path else f"docs/{fn}"

        if not link.is_symlink():
            print(f"  [skip]     {link.relative_to(repo)}  (not a symlink)")
            continue

        if not wiki_doc.exists():
            print(f"  [skip]     {display}  (wiki doc missing)")
            continue

        # Back up wiki doc before replacing with stripped version
        if scope is not None:
            local_edits_dir.mkdir(parents=True, exist_ok=True)
            backup_name = (rel_path.replace("/", "-") or "root") + ".md"
            shutil.copy(wiki_doc, local_edits_dir / backup_name)

        content = strip_wiki_header(wiki_doc.read_text())
        link.unlink()
        link.write_text(content)
        _no_skip_worktree(repo, str(link.relative_to(repo)))

        # Remove the flat mirror symlink from .agent-wiki/symlinks/
        flat = ("root" if not rel_path else rel_path.replace("/", "-")) + Path(fn).suffix
        mirror = symlinks_dir / flat
        if mirror.is_symlink():
            mirror.unlink()

        ejected_paths.append(rel_path)
        print(f"  [ejected]  {link.relative_to(repo)}")

    # Always clear conflict log for all requested paths — eject resolves any conflict
    all_requested = [rp for rp, _ in nodes]
    clear_conflict_log_for(all_requested)
    if not load_conflict_log():
        clear_flag("multiple_versions")

    ejected = len(ejected_paths)
    if ejected:
        print(f"\n{ejected} file(s) ejected. Wiki docs in docs/ are untouched.")
        print("To stop managing these paths, remove them from schema.yaml.")
    else:
        print("Nothing ejected.")

    if scope is None:
        _backup_docs()
        _remove_wiki_integration(repo)


def _backup_docs():
    if not DOCS_ROOT.exists():
        return
    bak = WIKI_ROOT / "docs.bak"
    if bak.exists():
        shutil.rmtree(bak)
    shutil.copytree(DOCS_ROOT, bak)
    print(f"  [backup]   docs/ → docs.bak/")


def _remove_wiki_integration(repo: Path):
    # .agent-wiki/ folder
    cw_dir = repo / ".agent-wiki"
    if cw_dir.exists():
        shutil.rmtree(cw_dir)
        print(f"  [removed]  .agent-wiki/")

    # Git hooks owned by agent-wiki
    for hook_name in ("pre-commit", "post-checkout"):
        hook = repo / ".git" / "hooks" / hook_name
        if hook.exists() and "agent-wiki" in hook.read_text():
            bak = hook.with_suffix(".agent-wiki.bak")
            hook.rename(bak)
            print(f"  [backup]   .git/hooks/{hook_name} → {bak.name}")

File: src/commands/eject.rs
use anyhow::{bail, Result};
use std::fs;

use crate::{
    config,
    git,
    logs,
    metadata,
    paths::{doc_path, symlink_path},
    root::WikiCtx,
    schema,
};

pub fn run(ctx: &WikiCtx, scope: Option<&str>, purge: bool) -> Result<()> {
    let cfg = config::load(&ctx.wiki_root)?;
    let sch = schema::load(&ctx.wiki_root)?;
    let mut managed = schema::walk(&sch, "");

    if let Some(s) = scope {
        if !managed.contains(&s.to_string()) {
            bail!(
                "'{}' is not a managed doc path.\nManaged paths: {}",
                s,
                managed.join(", ")
            );
        }
        managed.retain(|p| p == s);
    }

    let fn_ = &cfg.doc_filename;
    let repo = ctx.repo_root();
    let local_edits = ctx.wiki_root.join("logs").join("local-edits");
    let mut ejected = Vec::new();

    for rel_path in &managed {
        let link = symlink_path(repo, rel_path, fn_);
        let wiki_doc = doc_path(ctx, rel_path, fn_);

        if !link.is_symlink() {
            println!(
                "  [skip]     {}  (not a symlink)",
                link.strip_prefix(repo).unwrap_or(&link).display()
            );
            continue;
        }
        if !wiki_doc.exists() {
            println!(
                "  [skip]     {}  (wiki doc missing)",
                display_doc(rel_path, fn_)
            );
            continue;
        }

        // Scoped eject backs up to logs/local-edits/
        if scope.is_some() {
            fs::create_dir_all(&local_edits)?;
            let name = (if rel_path.is_empty() { "root" } else { rel_path }).replace('/', "-") + ".md";
            fs::copy(&wiki_doc, local_edits.join(&name))?;
        }

        let content = metadata::strip_header(&fs::read_to_string(&wiki_doc)?);
        fs::remove_file(&link)?;
        fs::write(&link, content)?;
        git::no_skip_worktree(repo, &link.strip_prefix(repo).unwrap().to_string_lossy());
        ejected.push(rel_path.clone());
        println!(
            "  [ejected]  {}",
            link.strip_prefix(repo).unwrap_or(&link).display()
        );
    }

    let all_paths: Vec<&str> = managed.iter().map(String::as_str).collect();
    logs::clear_conflicts_for(&ctx.wiki_root, &all_paths);
    if logs::load_conflicts(&ctx.wiki_root).is_empty() {
        logs::clear_flags(&ctx.wiki_root, &["multiple_versions"]);
    }

    if ejected.is_empty() {
        println!("Nothing ejected.");
    } else {
        println!(
            "\n{} file(s) ejected. Wiki docs in .agent-wiki/docs/ are untouched.",
            ejected.len()
        );
        println!("To stop managing these paths, remove them from schema.yaml.");
    }

    if scope.is_none() {
        remove_wiki_integration(ctx, purge)?;
    }

    Ok(())
}

fn display_doc(rel_path: &str, fn_: &str) -> String {
    if rel_path.is_empty() {
        format!("docs/{fn_}")
    } else {
        format!("docs/{rel_path}/{fn_}")
    }
}

fn remove_wiki_integration(ctx: &WikiCtx, purge: bool) -> Result<()> {
    let repo = ctx.repo_root();

    // Back up git hooks owned by agent-wiki
    for hook_name in ["pre-commit", "post-checkout"] {
        let hook = repo.join(".git").join("hooks").join(hook_name);
        if hook.exists() {
            let content = fs::read_to_string(&hook).unwrap_or_default();
            if content.contains("agent-wiki") {
                let bak = hook.with_extension("agent-wiki.bak");
                fs::rename(&hook, &bak)?;
                println!("  [backup]   .git/hooks/{hook_name} → {}", bak.file_name().unwrap_or_default().to_string_lossy());
            }
        }
    }

    if purge {
        // Remove .agent-wiki/ entirely — docs are in git history
        if ctx.wiki_root.exists() {
            fs::remove_dir_all(&ctx.wiki_root)?;
            println!("  [removed]  .agent-wiki/  (docs preserved in git history)");
            println!("  Recover:   git clone <wiki-remote> .agent-wiki/ && agent-wiki push --verify");
        }
    } else {
        println!("\n  .agent-wiki/ preserved (docs remain in git history).");
        println!("  To fully remove: agent-wiki eject --purge");
    }

    Ok(())
}
